src/alvin0319/OffHand/OffHand.php

```diff
@@ -30,84 +30,88 @@
-		if($player->namedtag->hasTag("OffHand", CompoundTag::class)){
-			$this->getOffHandInventory($player)->setItemInOffHand(Item::nbtDeserialize($player->namedtag->getCompoundTag("OffHand")));
-		}
-	}
-
-	private function saveInventory(Player $player) : void{
-		$player->namedtag->setTag($this->getOffHandInventory($player)->getItemInOffHand()->nbtSerialize(-1, "OffHand"));
-		unset($this->inventories[$player->getName()]);
+        	$player = $event->getPlayer();
+        	$this->getOffhandInventory($player)->sendContents($player);
 	}
 
+	/**
+	 * @ignoreCancelled true
+	 * @priority MONITOR
+	 */
 	public function onDataPacketReceive(DataPacketReceiveEvent $event) : void{
 		$packet = $event->getPacket();
 		$player = $event->getPlayer();
-		if($packet instanceof MobEquipmentPacket){
-			if($packet->windowId === ContainerIds::OFFHAND){
-				$inv = $this->getOffHandInventory($player);
-				if($inv instanceof OffHandInventory){
-					$inv->setItemInOffHand($packet->item);
-					$event->setCancelled();
-				}
-			}
-		}
+        	if($packet instanceof MobEquipmentPacket and $packet->windowId == ContainerIds::OFFHAND){
+            		$offhand = $this->getOffhandInventory($player);
+            		if(!$offhand->getItem(0)->equalsExact($packet->item)){
+                		$offhand->sendContents($player);
+                		$event->setCancelled();
+                		return;
+            		}
+            		$offhand->setItem(0, $packet->item);
+        	}
 	}
-
+	/**
+	 * @ignoreCancelled true
+	 * @priority MONITOR
+	 */
 	public function onDataPacketSend(DataPacketSendEvent $event) : void{
-		$packet = $event->getPacket();
-		if($packet instanceof AddPlayerPacket){
-			if(($player = $this->getServer()->getPlayerExact($packet->username)) !== null){
-				$this->getOffHandInventory($event->getPlayer())->sendMobEquipmentPacket($player);
-				$this->getOffHandInventory($event->getPlayer())->sendMobEquipmentPacket($event->getPlayer());
-			}
-		}
+        	$packet = $event->getPacket();
+        	$player = $event->getPlayer();
+        	if($packet instanceof AddPlayerPacket){
+            		$this->getOffhandInventory($player->getServer()->getPlayerExact($packet->username) ?? $player)->sendItem([$player]);
+        	}
 	}
-
+	/**
+	 * @priority MONITOR
+	 */
 	public function onDeath(PlayerDeathEvent $event) : void{
-		$player = $event->getPlayer();
-		$drops = $event->getDrops();
-		if(!$event->getKeepInventory()){
-			$drops = array_merge($drops, $this->getOffHandInventory($player)->getContents(false));
-			$event->setDrops($drops);
-			$this->getOffHandInventory($player)->clearAll();
+        	$player = $event->getPlayer();
+        	$offhand = $this->getOffhandInventory($player);
+        	if(!$event->getKeepInventory() and !empty($event->getDrops())){
+            		$event->setDrops(array_merge($event->getDrops(), array_filter([$offhand->getItem(0)], function(Item $item): bool{
+                		return !$item->hasEnchantment(Enchantment::VANISHING);
```

Comment:
What is this purpose?

Reply:
Oh that if the item in the offhand has the enchantment of vanishing and if the drops in the Death event are empty then is because MultiWorld has enabled keep inventory

Reply:
You know, maybe in the future you are able to use any item in offhand, so a item could have that enchant

Reply:
Does PMMP implemented vanish enchantment?